pkg/manifest/pipeline.go

```diff
@@ -174,7 +174,9 @@ func NewBase(name string, build Build) Base {
 
 // serializeStart must be called exactly once before each call
 // to serialize().
-func (p Base) serializeStart(inputs Inputs) {
+func (p Base) serializeStart(inputs Inputs) error {
+	// XXX: we could do the "double call" check and other common mistakes here
```

Comment:
Have a consistent way of doing this here would actually be nice. I ran into issues with the current "indirect" way of tracking via things like containerSpecs and its also from a casual reading view not obvious that the set/unset of eg. containerSpecs has this side effect so being explicit here might be a good idea.

Reply:
I had the same thought.

My initial idea would be to add an explicit `bool` variable in the `Base` to track this. And maybe doing the checks in the `Base.serializeStart()` and then all structs that embed it would call it as the first thing in their `serializeStart()` implementation.